myokit/formats/sbml/_importer.py

```diff
@@ -35,334 +36,1065 @@ def __init__(self):
-        x = xmodel.find(self.ns['sbml'] + 'listOfConstraints')
-        if x:   # pragma: no cover
-            log.warn('Constraints are not supported.')
-        x = xmodel.find(self.ns['sbml'] + 'listOfReactions')
-        if x:   # pragma: no cover
-            log.warn('Reactions are not supported.')
-        x = xmodel.find(self.ns['sbml'] + 'listOfEvents')
-        if x:   # pragma: no cover
-            log.warn('Events are not supported.')
-
-        # Ignore custom functions
-        x = xmodel.find(self.ns['sbml'] + 'listOfFunctionDefinitions')
-        if x:   # pragma: no cover
-            log.warn('Custom math functions are not (yet) implemented.')
-
-        # Parse custom units
-        x = xmodel.find(self.ns['sbml'] + 'listOfUnitDefinitions')
-        if x:
-            self._parse_units(model, comp, x)
-
-        # Add time as independent variable (not explicit in SBML format)
-        if bind_time:
-            # Add and bind time variable to component
-            time = comp.add_variable_allow_renaming('time')
-            time.set_binding('time')
-
-            # Set unit and value
+        # Raise error if function definitions are provided (could be added in
+        # another PR)
+        funcDefs = self._getListOfFunctionDefinitions(SBMLmodel)
+        if funcDefs:
+            raise SBMLError(
+                'Myokit does not support functionDefinitions. Please insert '
+                'your function wherever it occurs in yout SBML file and delete'
+                ' the functionDefiniton in the file.')
+
+        # Create user defined unit reference
+        self.userUnitDict = dict()
+
+        # Get unit definitions
+        unitDefs = self._getListOfUnitDefinitions(SBMLmodel)
+        if unitDefs:
+            for unitDef in unitDefs:
+                unitId = unitDef.get('id')
+                if not unitId:
+                    raise SBMLError(
+                        'The file does not adhere to SBML 3.2 standards.'
+                        ' No unit ID provided.')
+                self.userUnitDict[
+                    unitId] = self._convert_unit_def(unitDef)
+
+        # Get model units
+        units = {
+            'substanceUnit': SBMLmodel.get('substanceUnits'),
+            'timeUnit': SBMLmodel.get('timeUnits'),
+            'volumeUnit': SBMLmodel.get('volumeUnits'),
+            'areaUnit': SBMLmodel.get('areaUnits'),
+            'lengthUnit': SBMLmodel.get('lengthUnits'),
+            'extentUnit': SBMLmodel.get('extentUnits'),
+        }
+        for unitId in units:
+            unit = units[unitId]
+            if unit in self.userUnitDict:
+                self.userUnitDict[unitId] = self.userUnitDict[unit]
+            elif unit in SBML2MyoKitUnitDict:
+                self.userUnitDict[unitId] = SBML2MyoKitUnitDict[unit]
+            else:
+                self.userUnitDict[unitId] = None
+
+        # Initialise parameter and species dictionary
```

Comment:
I assume the contents of these dicts are quite specific? As in, they need to have some specific structure?
In that case, please either add a bunch of comments explaining this structure (could be as simple as "maps ids to myokit.Component objects"), or alternatively create some class structure for this information to live in